### DISCUSSION TITLE
conditional settings not working on comands config for tox settings

### DISCUSSION THREAD

**Discussion by rahulmukherjee68** [tox-dev/tox#2458] [General]
When submitting a bug make sure you can reproduce it via ``tox -rvv`` and attach the output of that to the bug. Ideally, you should also submit a project that allows easily reproducing the bug. Thanks!

The below is subpart of my tox ini when i am applying conditional factor to run a particular command its causing invocation error 
```
[tox]
envlist=
    {test,e2e,codechecker}-{py36,py37,py38,py39}-{oracledb,mysql,pymysql}

commands =
    test-!oracledb: pip install -r {toxinidir}/test-requirements.txt
```

Error:- `ERROR: InvocationError: could not find executable 'test-!oracledb:'`

**I want to run the pip install command for all the cases execpt when there is oracledb in my factor can you please assist me if there is any issue in the condition or its an internal issue**

**Comment by jugmac00** [comment]
```
commands =
    !oracledb: pip install -r {toxinidir}/dev-requirements.txt
```

.... should do the trick.

Also see https://tox.wiki/en/latest/config.html#complex-factor-conditions

**Reply by rahulmukherjee68** [reply]
Thank for giving the feedback
 The question is the i want to run the command when the test factor is there in env 
But if test as well as oracledb is there in env i don't want to run it.

There are other factors also in my env list i have not given here  so i have to add a test factor also in ur above condition can you tell me how I can do it .. 

PS :- i want the same condition as of if there is test except oracledb

**Reply by jugmac00** [reply]
Thanks for clarifying as this information was not available in your question from the beginning.

I think it would sense to provide a more complete example.

Your envlist looks like this:

```
envlist=
    {test,e2e,codechecker}-{py36,py37,py38,py39}-{oracledb}
```

which expands to...

```
❯ tox -a
test-py36-oracledb
test-py37-oracledb
test-py38-oracledb
test-py39-oracledb
e2e-py36-oracledb
e2e-py37-oracledb
e2e-py38-oracledb
e2e-py39-oracledb
codechecker-py36-oracledb
codechecker-py37-oracledb
codechecker-py38-oracledb
codechecker-py39-oracledb
```

**Reply by rahulmukherjee68** [reply]
Yes the the environment looks like this but i have more factor in list  of db which results in multiple environment of db but when we have factors of  `test-oracledb` i want to skip the above mentioned command
But for rest db it should run like `test-py36-pymysql` for this pip install command should run

**Reply by jugmac00** [reply]
This should work:

```
test-py{36,37,38,39}-!oracledb
```

**Reply by rahulmukherjee68** [reply]
i have done that only in my summary it not working its giving invocation error

**Comment by jugmac00** [comment]
Rahul, thanks for providing all the necessary information.

This should work:

```
test-py{36,37,38,39}-!oracledb
```

**Reply by rahulmukherjee68** [reply]
yes i tried it but its not working its giving the same invocation error . can you try in you env i am using tox==2.7.0

**Reply by jugmac00** [reply]
The above solution works with a recent, and supported version of tox.

Please understand that we do not offer support for versions which are outdated for 5 years.

While - again - you do not tell me the whole story, I assume you have installed tox via your operating system's package manager. This is usually ok, but only as long as that specific version supports the features that you need.

Either get support from your distribution of choice or install a supported version of tox.

For how to install a newer version of tox, please see https://tox.wiki/en/latest/install.html

I won't be able to help you any further with this issue.

If you don't mind... next time when you ask a question, please include all the requested and necessary data, otherwise it is very hard, cumbersome or even impossible to help you.

**Reply by rahulmukherjee68** [reply]
Sure . Thank you for your feedback would be taking in account for the full summary statement with necessary data if any discussion comes up in future issues. Thanks for helping would be useful for me
